crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,683 @@
+        }
+        // check if the block is already present in the soft blocks
+        if let Some(soft_block) = self.soft_blocks.get(&height) {
+            if soft_block.block_hash() == block_hash {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Check if there is another block in the pending blocks at a lower height
+    /// that points to this block as its parent.
+    fn is_block_a_parent(&mut self, block: SequencerBlockSubset) -> bool {
+        let block_hash = block.block_hash();
+        if let Some(child_blocks) = self.pending_blocks.get(&block.child_height()) {
+            let blocks = child_blocks.values();
+            for block in blocks {
+                if let Some(hash) = block.parent_hash() {
+                    if hash == block_hash {
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Safely insert a block into the pending blocks. This function will insert
+    /// the block into existing maps or create a new one if needed.
+    fn insert_to_pending_blocks(&mut self, block: SequencerBlockSubset) {
+        let height = block.height();
+        let block_hash = block.block_hash();
+
+        if let Some(pending_blocks) = self.pending_blocks.get_mut(&height) {
+            let mut new_map = pending_blocks.clone();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        }
+        self.update_internal_state();
+    }
+
+    // remove all data in the queue below a given height. this does not remove
+    // data from the soft queue, only the pending queue, and updates the head height.
+    // TODO: add error handling
+    fn remove_data_below_height(&mut self, height: Height) {
+        // remove all data below the new incoming block from the pending data
+        let tmp_pending = self.pending_blocks.clone();
+        let mut pending_keys: Vec<&Height> = tmp_pending.keys().collect();
+        pending_keys.sort();
+        for key in pending_keys {
+            if *key < height {
+                self.pending_blocks.remove(key);
+            }
+        }
+    }
+
+    /// This function organizes the internal state of the queue based on the
+    /// tendermint/CometBTF fork choice rules.
+    ///
+    /// Once a block is added to the pending_blocks in the queue, this function
+    /// is called. It walks the pending blocks from lowest to highest height,
+    /// checking to see if there is a continuous chain of blocks. For every block
+    /// that is a descendant of the most recent "soft" block, and has a direct
+    /// descendant, that block gets added to the `soft_blocks` BTreeMap and the
+    /// head height is updated.
+    fn update_internal_state(&mut self) {
```

Comment:
This function performs reallocation of a two dimensional data structure on potentially every iteration of a loop.

Furthermore, adding new blocks to the `pending_blocks` queue probably happens very frequently, so this function does a lot of unnecessary work.

You can fix this by making use of a `loop {}` instead of iterating over the heights directly (and thus holding on to a reference of into `self`, making it impossible to also mutate it.

First I think it's necessary to change `pending_blocks` from a `HashMap` to a `BTreeMap` so that it's self-ordering (which IMO makes sense because you almost always collect + sort it anyway). Then you can make use of the fact that you are effectively using `head_height` as the iteration index.

Then use `loop {}`, using `head_height` to index into `self.pending_blocks`, checking if you have a parent, mutating the inner state, etc. Finally, increment `head_height` and define a stopping condition (for example, `head_height > self.pending_blocks.last_key_value.0`).